### PR TITLE
[data, trainer] feat: add text training-time validation

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

> Add training-time validation for text SFT without carrying forward the unsafe cross-trainer semantics from closed PR #1077.
>
> The initial scope is intentionally narrow: single-source, map-style conversation or classification data with the native dataloader under pure DP/FSDP2. Unsupported trainer, modality, data, parallel, and observability modes fail closed.

### Checklist Before Starting

- Search for relative PRs/issues and link here: replaces closed #1077. A live open-PR scan found path overlap with #936, #994, #1076, and #1082, but no open PR implementing validation or `eval_path` semantics.
- PR title follows `[{modules}] {type}: {description}` format.

### Test

> Passed on Python 3.12.13 / PyTorch 2.11.0:
>
> - `PYTHONPATH=. python -m pytest -q tests/data/test_validation_sampler.py tests/trainer/test_training_validation.py` — 46 passed, including a real two-rank Gloo reduction with uneven local batch sizes and causal-target counting regressions for valid first labels and one-token sequences.
> - `PYTHONPATH=. python -m pytest -q tests/data/test_dataloader.py` — 22 passed.
> - `PYTHONPATH=. python -m pytest -q tests/data/test_collators.py tests/data/test_classification_data_processor.py` — 15 passed.
> - `make quality` and `git diff --check` passed.
>
> Two-GPU FSDP2 functional smoke passed on the exact current PR head `5ebf719d7e9c8e97f5bc0abdadc81cc3c72b27e6` against `main@cf5dd272d616fc2ae14025225dc036948e6440ae`:
>
> - 2× NVIDIA A800 80GB, one node, two-rank NCCL, pure DP with FSDP2 shard degree 2; all other parallel dimensions were 1.
> - A randomly initialized two-layer Qwen2 toy model ran three BF16 AdamW optimizer updates over deterministic synthetic map-style data (6 train / 4 validation samples, sequence length 128, seed 1083).
> - Training losses were `12.279469`, `12.272757`, and `12.123192`; gradient norms were `10.614016`, `10.838994`, and `11.004934`.
> - The two-rank NCCL preflight returned the expected all-reduce sum `3.0`.
> - Validation ran exactly at global steps 2 and 3 with losses `12.261265` and `12.258946`; each pass aggregated exactly 508 shifted causal targets, restored all module modes, and returned the model to training mode.
> - Slurm job `310084` completed with exit code 0. The runtime was Python 3.12.3, PyTorch 2.10.0+cu129, CUDA 12.9, and Transformers 5.9.0.
> - This validates accelerator batch transfer, FSDP2 forward execution, NCCL collectives, weighted distributed validation reduction, callback scheduling, and module-mode restoration on the current PR head.

### API and Usage Example

```yaml
data:
  eval_path: /path/to/validation.jsonl
  datasets_type: mapping
train:
  eval_steps: 100
  eval_epochs: 0
```

> `eval_path: null` disables training-time validation. See `docs/usage/trainer.md` for the complete support and rejection matrix.

### Design & Code Changes

> - Let `TextTrainer` own a dedicated validation runner while the callback only schedules it; other composed trainers reject validation until they provide task-specific contracts.
> - Add an O(1) exact distributed batch sampler that visits every record once, preserves source order, keeps equal non-empty forward counts across ranks, and rejects impossible partitions instead of padding or duplicating data.
> - Aggregate causal loss by shifted non-ignored tokens and classification loss by non-ignored sample labels over the captured DP group.
> - Restore every module's prior train/eval mode on success or failure, isolate dataloader RNG, and disable persistent validation workers.
> - Reject plaintext chunk expansion, weighted multisource YAML, iterable data, non-text trainers, SP/TP/PP/EP, async SP, compile, profiler, and MoE router monitoring until their contracts are implemented.
> - Add property, fault-injection, real collator/dataloader, cross-trainer rejection, and two-process collective tests; document the public configuration boundary.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- Applied scoped pre-commit checks; the unrelated repository-wide formatting baseline is documented above.
- Added/updated documentation.
- Added CPU tests for sampler, runner, callback, collator/dataloader integration, and distributed aggregation; completed the two-GPU FSDP2 functional smoke documented above.

## Summary by CodeRabbit

* **New Features**
  * Added training-time validation for text SFT workflows.
  * Validation reports loss weighted by non-ignored targets and supports distributed evaluation.
  * Added deterministic, non-padding validation batching with equal step distribution.
  * Validation preserves model state and restores it after evaluation.
  * Validation metrics can be logged through evaluation callbacks and Weights & Biases.

* **Bug Fixes**
  * Added clear errors for unsupported validation configurations and invalid batches.

* **Documentation**
  * Documented validation setup, behavior, supported scope, limitations, and configuration options.
  * Clarified how auxiliary metrics are handled during training.

### CODE REVIEW

File: veomni/trainer/validation.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            **dataloader_kwargs,
+        )
+        return dataset, dataloader
+
+    def _count_loss_units(self, labels: torch.Tensor) -> torch.Tensor:
+        if self.trainer.args.data.data_type == "classification":
+            return torch.count_nonzero(labels != IGNORE_INDEX)
+        if labels.ndim == 0 or labels.shape[-1] < 2:
+            return labels.new_zeros((), dtype=torch.long)
+        return torch.count_nonzero(labels[..., 1:] != IGNORE_INDEX)
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/trainer/validation.py:165]: Thanks. I checked this against the causal-LM loss contract: logits[..., :-1] predict labels[..., 1:], so labels[..., 0] has no preceding logit and must not contribute even when it is valid; a one-token causal sample therefore has zero loss targets. `count_loss_token` serves different training bookkeeping and is not the correct denominator for the model-reported causal loss. Commit 9d4e0d19 adds regressions for a valid first label, one-token input, and ignore-mask placements; the combined validation suite passes 46 tests. I am keeping the implementation unchanged and treating this as a false positive.

**Comment by WenzheWang** [comment]
@Coach257 This PR is synchronized with the current `main` base (`cf5dd272`; current head `5ebf719d`). The focused 46-test validation suite and `make quality` pass, and the exact current head has now also passed a 2×A800 NCCL/FSDP2 functional smoke (Slurm job `310084`). The PR body contains the full current-head setup and results. Could you please review the updated PR and approve the required workflows when convenient? Thanks.

**Comment by WenzheWang** [comment]
@coderabbitai I validated the current PR head `5ebf719d7e9c8e97f5bc0abdadc81cc3c72b27e6` against the accelerator-backed FSDP2 review request.

- The automated two-rank Gloo test covers uneven exact-batch distribution and weighted reduction.
- Slurm job `310084` ran on 2× NVIDIA A800 80GB with NCCL and pure DP/FSDP2 shard degree 2.
- The exact-head run completed three finite BF16 optimizer updates and the NCCL preflight returned the expected all-reduce sum `3.0`.
- Validation ran at global steps 2 and 3 with losses `12.261265` and `12.258946`, aggregating 508 shifted causal targets per pass.
- All module modes were restored after each validation pass and the model returned to training mode.

The PR body is updated with the exact-head configuration and results. Together, the automated distributed test and exact-head accelerator smoke cover uneven exact batching, accelerator transfer, FSDP2 forward execution, NCCL collectives, distributed reduction, scheduling, and state restoration.

**Comment by WenzheWang** [comment]
Closing after 35 days of inactivity. The text training-time validation feature has been superseded by later work on the main branch. Thanks to @Coach257 and @coderabbitai for their reviews.
